### DISCUSSION TITLE
How to backtest a machine learning model which uses a sequence data as input?

### DISCUSSION THREAD

**Discussion by AboveTheHeavens** [kernc/backtesting.py#339] [Q&A] [answered]
I have an ML model which uses a 128 value period (think 128 rows) to predict 128+1's output value.

I have tried to look at documentation but from my understanding I have to input a df with open, high, low, close columns.

How would I go about passing my 128 row sequence input?

Thanks a lot!

**Comment by kernc** [comment, marked as answer]
Have you seen ["Trading with Machine Learning" tutorial](https://kernc.github.io/backtesting.py/doc/examples/Trading%20with%20Machine%20Learning.html)? It shows an example of a walk-forward analysis.

Basically, you pass the whole `df` into `Backtest`; then `df.iloc[:128]` in `init()` (or `df.iloc[-128:]` in `next()`) at each model training stage.

**Reply by AboveTheHeavens** [reply]
Wow thanks a lot!

**Reply by AboveTheHeavens** [reply]
I don't think this works, if I try to print `print(self.data.df.shape)`, it gives me `(2, 6)`, so I'm only getting 2 rows at a time in next?

**Reply by kernc** [reply]
Well, `next()` ["builds up" data incrementally](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.data&gsc.tab=0). Have you read ["Trading with Machine Learning" tutorial](https://kernc.github.io/backtesting.py/doc/examples/Trading%20with%20Machine%20Learning.html)? It contains:
```py
    def next(self):
        # Skip the training, in-sample data
        if len(self.data) < N_TRAIN:
            return
```
In the same way, you can skip a few steps until sufficient data is available?
